### ISSUE TITLE
Testing and fixing bugs

### ISSUE THREAD

**Issue by jan-petr** [ExploreASL/ExploreASL#483]
### RELEASE NOTES #483
* Fixes of bugs introduced by #290
* `xASL_ut_UnitTest_function_BIDS2Legacy`: Minor improvements
* `xASL_im_dilateROI`: Fix bug (where was this introduced, when the new argument was added?)
* `xASL_tsvRead`: Fix for numerical arrays
* `xASL_bids_BIDS2Legacy`: BIDS import fixes
* `xASL_SysCopy`: Fix directory copy for Linux
* `xASL_bids_CompareStructures`: Minor update NIFTI comparison

### Initial comments by Jan Petr:

We can rename this issue later. But I wanted to test everything and fix all the major bugs after the recent merge of several branches.

Please list the bugs as part of this issue.

1. (FIXED) Debugging Full_BIDS_Pipeline_Test 
2. (FIXED) xASL_init_FileSystem.m - line 39 crashes at xASL_adm_CreateDir(x.D.T1CheckDir);  as the value is unknown
3. (FIXED) Incorrect header of ExploreASL_Master
4. (FIXED) Calls of ExploreASL_Master were not updated to the new version throughout
5. (FIXED) xASL_Iteration - line 376 - incoherence in variable name
6. (FIXED) Philips_2DEPI_Lesion_NoFLAIR_NoM0_HiQ - crashes in Structural module - Last step structural model in dilation lesions (see the image below in the comments). The error was an incorrect call of xASL_im_dilateROI - the function call needs to be updated everywhere
7. Error in processing flavors 
  - (FIXED) Philips_PCASL_2DEPI_Ingenia_volunteer, 
  - (FIXED) Philips_PCASL_2DEPI_dummyLL, 
  - (FIXED) Philips_PCASL_2DEPI_dummyMultiPLD, 
  - (FIXED) Philips_PCASL_2DEPI_dummyQUASAR, 
  - (FIXED) Philips_PCASL_3DGRASE_R5.4_TopUp
  - (FIXED) Philips_PCASL_3DGRASE_functional
  - (FIXED) Siemens_PASL_2DEPI_noBsup_AD, Siemens_PASL_2DEPI_noBsup_AD2 - M0 problem only
  - Siemens_PASL_3DGRASE_Prisma_Bsup_AD1
  - Siemens_PASL_3DGRASE_Prisma_Bsup_AD3
  - (FIXED) Siemens_PCASL_multiTI
  - Siemens_PCASL_volunteer

**Comment by MichaelStritt** [comment]
Just documentation, but in `ExploreASL_Master` `% FORMAT: [x] = ExploreASL([DataParPath, ImportModules, ProcessModules, SkipPause, iWorker, nWorkers])` should be `% FORMAT: [x] = ExploreASL([DataParPath, ImportModules, ProcessModules, bPause, iWorker, nWorkers])` instead.

**Comment by MDijsselhof** [comment]
**Error in the structural module**


![WhatsApp Image 2021-04-09 at 15 12 22](https://user-images.githubusercontent.com/75380250/114187047-dc23e600-9947-11eb-8913-5ffcf14275bd.jpeg)

**Origin of error, with pathIn shown**


![WhatsApp Image 2021-04-09 at 12 06 22](https://user-images.githubusercontent.com/75380250/114187127-f231a680-9947-11eb-9d4f-b47e5aa2b3f1.jpeg)

**Comment by MichaelStritt** [comment]
Is that related to the unix path issue?

**Comment by jan-petr** [comment]
> Just documentation, but in `ExploreASL_Master`...
> `% FORMAT: [x] = ExploreASL([DataParPath, ImportModules, ProcessModules, SkipPause, iWorker, nWorkers])`
> ...should be...
> `% FORMAT: [x] = ExploreASL([DataParPath, ImportModules, ProcessModules, bPause, iWorker, nWorkers])` instead.

Done

**Comment by jan-petr** [comment]
> Is that related to the unix path issue?

No - that was there probably before. Because the full path is given. I'm investigating it, but have to fix all the errors I have encountered before that :D

**Comment by MDijsselhof** [comment]
Apologies, the PathIn is not the problem I think but PathOut. This is set to 40 for some reason

**Comment by MichaelStritt** [comment]
Ah sorry, I changed this part in one of the last commits before the merge to remove the path from the json. Nice fix 👍

**Comment by jan-petr** [comment]
> Ah sorry, I changed this part in one of the last commits before the merge to remove the path from the json. Nice fix

No worries - that was an easy fix :)

**Comment by jan-petr** [comment]
![image](https://user-images.githubusercontent.com/29886537/114224307-35543f80-9971-11eb-908d-97f3d61b1394.png)

Latest error - all above fixed.

**Comment by MichaelStritt** [comment]
We should also fix the format definition in the header for `[x] = ExploreASL(varargin)`

**Comment by MichaelStritt** [comment]
### Test using the first two GE Flavors

@HenkMutsaerts & @jan-petr: I needed to change this right here 8da72af. I don't get how it worked for Jan if he didn't change that part though? There were also quite a few warnings. Check out the files below:

[logContent.xlsx](https://github.com/ExploreASL/ExploreASL/files/6291720/logContent.xlsx) & [fullPipelineTest.txt](https://github.com/ExploreASL/ExploreASL/files/6291723/fullPipelineTest.txt)

**Comment by MichaelStritt** [comment]
### Additional question

In `xASL_test_BIDSFlavorsFull` line **176**: if we don't want to run the population module, we should use `[1 1 0] `for the `ProcessModules` or do we want to run the population module now?

**Comment by HenkMutsaerts** [comment]
> ### Additional question
> In `xASL_test_BIDSFlavorsFull` line **176**: if we don't want to run the population module, we should use `[1 1 0] `for the `ProcessModules` or do we want to run the population module now?

You could but it is a bit overkill. We should definitely test the behavior of the population module as well in several conditions, but this is more dependent on image quality and pipeline parameters, which are tested in `TestDataSets`/`xASL_qc_TestExploreASL`. Especially the ASL module is dependent on flavor differences, and the structural module a bit. 

So I would indeed disable the population module in `xASL_test_BIDSFlavorsFull`

**Comment by MichaelStritt** [comment]
Will do 👍

**Comment by HenkMutsaerts** [comment]
> ### Test using the first two GE Flavors
> @HenkMutsaerts & @jan-petr: I needed to change this right here [8da72af](https://github.com/ExploreASL/ExploreASL/commit/8da72afa3ce8ad151ee0ac91a8d3be46a73d40a5). I don't get how it worked for Jan if he didn't change that part though? 

You are right, perhaps Jan corrected the incorrect copied path in a later stage? The whole temporary copying is unnecessary right, and potentially adds bugs? It is more important that a check is being done on the "read-only-ness" of the `sourcedata` folder (otherwise issuing a warning that this should be set to read-only for good practice).

The source data should always be read-only. The rawdata could be set to read-only after the conversion, the derivatives can be anything.

**Comment by MichaelStritt** [comment]
> 
> 
> > ### Test using the first two GE Flavors
> > @HenkMutsaerts & @jan-petr: I needed to change this right here [8da72af](https://github.com/ExploreASL/ExploreASL/commit/8da72afa3ce8ad151ee0ac91a8d3be46a73d40a5). I don't get how it worked for Jan if he didn't change that part though?
> 
> You are right, perhaps Jan corrected the incorrect copied path in a later stage? The whole temporary copying is unnecessary right, and potentially adds bugs? It is more important that a check is being done on the "read-only-ness" of the `sourcedata` folder (otherwise issuing a warning that this should be set to read-only for good practice).
> 
> The source data should always be read-only. The rawdata could be set to read-only after the conversion, the derivatives can be anything.

There's one step in the comparison script where we open the gzipped niftis and compare the image values using RMSE. We should work on a nifti image reader which works on a copy, then this would be way much easier.

**Comment by MichaelStritt** [comment]
Similar change to the 'sourcedata' change from 8da72af here in 3e87fa3. Otherwise this doesn't work correctly for me on windows.

**Comment by HenkMutsaerts** [comment]
>> Please note the warnings in the txt file, which are related to the comparison script.

Is this your first time testing this script? :)

So: 
* you can ignore the population module errors/warnings (per what I suggested below, to skip this module)
* Jan did not add a structural scan to each ASL scan, and its correct that the pipeline gives a warning in this case, you can ignore this warning
* The warning on the `notification handle`, is this a permission issue?
* The other warnings concern conversion stuff that Jan only can answer :)

**Comment by MichaelStritt** [comment]
> The warning on the notification handle, is this a permission issue?

Yep, you can ignore those.

> The other warnings concern conversion stuff that Jan only can answer :)

Should be fixed within 3e87fa3. Hope it still works on linux though.

**Comment by HenkMutsaerts** [comment]
>> There's one step in the comparison script where we open the gzipped niftis and compare the image values using RMSE. We should work on a nifti image reader which works on a copy, then this would be way much easier.

Opening read-only NIfTIs should not be an issue, though the unzipping of a .nii.gz occurs locally, which would then not be possible inside `rawdata`. We could indeed specify a working directory that is writeable, and when opening a NIfTI, let `xASL_io_ReadNifti` unzip in the working directory instead of in the local folder (if the local folder is read-only).

So I would suggest to create the following issues:
1. Add serious warning to BIDS import if sourcedata is not read-only
2. include try-catch statement in `xASL_io_ReadNifti` if it cannot unzip a .nii.gz in a local read-only folder (such as 'rawdata') and unzip in another temporary folder.

**Comment by HenkMutsaerts** [comment]
> 3e87fa3

Looks good

**Comment by MichaelStritt** [comment]
> Opening read-only NIfTIs should not be an issue, though the unzipping of a .nii.gz occurs locally, which would then not be possible inside rawdata. We could indeed specify a working directory that is writeable, and when opening a NIfTI, let xASL_io_ReadNifti unzip in the working directory instead of in the local folder (if the local folder is read-only).

Don't think it's a read/write issue. Only the unzipping part. If you do this within the FlavorDatabase repository, you'd have to reset afterwards. I would add an option to xASL_io_ReadNifti to unzip a copy (or just unzip without deleting the original) that deletes the unzipped version afterwards. Or we just xASL_Copy the Nifti in the comparison script, then unzip it and use xASL_Delete afterwards. That's probably the easiest way.

> Wait, let me check if the behavior of xASL_SysCopy is correct, to be sure

Yep, there could be a bug there and we changed the behavior in the v1.5.1 release, right?

So from my side it looks quite good now. At least for the first flavor:
[logContent.xlsx](https://github.com/ExploreASL/ExploreASL/files/6291816/logContent.xlsx) & [fullPipelineTest.txt.txt](https://github.com/ExploreASL/ExploreASL/files/6291815/fullPipelineTest.txt.txt)

**Comment by HenkMutsaerts** [comment]
>> Don't think it's a read/write issue. Only the unzipping part. If you do this within the FlavorDatabase repository, you'd have to reset afterwards. I would add an option to xASL_io_ReadNifti to unzip a copy (or just unzip without deleting the original) that deletes the unzipped version afterwards. Or we just xASL_Copy the Nifti in the comparison script, then unzip it and use xASL_Delete afterwards. That's probably the easiest way.

OK, the point of the `xASL_` .nii.gz behavior is that you don't want a copy of a file that might be subtly changed. So it is better to delete the .nii.gz when unzipping to .nii (unless you would always delete the .nii afterwards and always save to .nii.gz but this would take up a lot of extra time when doing this for intermediate files).

When you add an option to `xASL_io_ReadNifti` to unzip without deleting the original .nii.gz, this should be allright as the next time this NIfTI pair is touched, both files will be compared and one deleted automatically, but this could potentially create an error.

It would be ideal to do the unzipping in memory, but this is not possible in combination with the memory mapping of NIfTI as a Matlab object, that we (and SPM) make use of a lot. Hopefully a next SPM release would allow memory mapping of a .nii.gz and unzipping in memory.

**Comment by HenkMutsaerts** [comment]
@jan-petr should we remove or catch import warnings that are flavor specific (e.g. for this GE example it says that it cannot find `ProtocolName` which I believe is a Philips-specific DICOM field?

**Comment by MichaelStritt** [comment]
> OK, the point of the xASL_ .nii.gz behavior is that you don't want a copy of a file that might be subtly changed. So it is better to delete the .nii.gz when unzipping to .nii (unless you would always delete the .nii afterwards and always save to .nii.gz but this would take up a lot of extra time when doing this for intermediate files).
> 
> When you add an option to xASL_io_ReadNifti to unzip without deleting the original .nii.gz, this should be allright as the next time this NIfTI pair is touched, both files will be compared and one deleted automatically, but this could potentially create an error.
> 
> It would be ideal to do the unzipping in memory, but this is not possible in combination with the memory mapping of NIfTI as a Matlab object, that we (and SPM) make use of a lot. Hopefully a next SPM release would allow memory mapping of a .nii.gz and unzipping in memory.

I chose the easier way and modified `xASL_bids_CompareStructures` to compare copies of the NIFTIs. We now could also compare the rawdata folders of the FlavorDatabase repo with the TmpConversion version. Check out e3dd0af.

**Comment by HenkMutsaerts** [comment]
> **Error in the structural module**
> 
> ![WhatsApp Image 2021-04-09 at 15 12 22](https://user-images.githubusercontent.com/75380250/114187047-dc23e600-9947-11eb-8913-5ffcf14275bd.jpeg)
> 
> **Origin of error, with pathIn shown**
> 
> ![WhatsApp Image 2021-04-09 at 12 06 22](https://user-images.githubusercontent.com/75380250/114187127-f231a680-9947-11eb-9d4f-b47e5aa2b3f1.jpeg)

So this bug needs to go to a separate issue?

**Comment by MDijsselhof** [comment]
> > **Error in the structural module**
> > ![WhatsApp Image 2021-04-09 at 15 12 22](https://user-images.githubusercontent.com/75380250/114187047-dc23e600-9947-11eb-8913-5ffcf14275bd.jpeg)
> > **Origin of error, with pathIn shown**
> > ![WhatsApp Image 2021-04-09 at 12 06 22](https://user-images.githubusercontent.com/75380250/114187127-f231a680-9947-11eb-9d4f-b47e5aa2b3f1.jpeg)
> 
> So this bug needs to go to a separate issue?

Jan has fixed this
